ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp

```diff
@@ -656,9 +664,103 @@ std::tuple<ttnn::Tensor, uint32_t, uint32_t, ttnn::Tensor, std::optional<ttnn::T
     uint32_t output_height = ((input_height - kernel_size[0] - ((kernel_size[0] - 1 ) * (dilation[0] - 1)) + 2 * padding[0]) / stride[0]) + 1;
     uint32_t output_width = ((input_width - kernel_size[1] - ((kernel_size[0] - 1 ) * (dilation[0] - 1)) + 2 * padding[1]) / stride[1]) + 1;
+
+    Conv2dConfig conv_config = conv_config_.value_or(Conv2dConfig());
+    if(conv_config.output_height_in_l1 > 0) {
+        TT_FATAL((conv_config.output_height_in_l1 % 32) == 0, "Output height in L1 must be a multiple of 32", conv_config.output_height_in_l1);
+        ttnn::Tensor input_tensor_on_device;
+        if( !is_tensor_on_device_or_multidevice(input_tensor)) {
+            input_tensor_on_device = ttnn::operations::core::to_device(input_tensor, device, std::nullopt);
+        }
+        else
+        {
+            input_tensor_on_device = input_tensor;
+        }
+        ttnn::Tensor weight_tensor_on_device;
+        std::optional<ttnn::Tensor> bias_tensor_on_device;
+        if(input_tensor_on_device.memory_config().is_dram()) {
+            Tensor dram_output_tensor;
+            bool first_run = true;
+            for(int batch_index = 0; batch_index < batch_size; batch_index++) {
+                for(uint32_t output_slice_height_start = 0; output_slice_height_start < output_height; output_slice_height_start+=conv_config.output_height_in_l1) {
+                    uint32_t output_slice_height_end = std::min(output_height, output_slice_height_start + conv_config.output_height_in_l1);
+                    uint32_t output_slice_height = output_slice_height_end - output_slice_height_start;
+
+                    if(output_slice_height == 0) {
+                        continue;
+                    }
+
+                    int input_slice_height_start = output_slice_height_start * stride[0] - padding[0];
+                    int input_slice_height_end = (output_slice_height_end - 1) * stride[0] - padding[0] + (kernel_size[0] - 1) * (dilation[0] - 1) + kernel_size[0];
+                    int pad_top = std::max(0, -input_slice_height_start);
+                    int pad_bottom = std::max<int>(0, input_slice_height_end - input_height);
+                    input_slice_height_start = std::max(0, input_slice_height_start);
+                    input_slice_height_end = std::min<int>(input_height, input_slice_height_end);
+                    uint32_t input_slice_height = input_slice_height_end - input_slice_height_start;
+                    log_debug(tt::LogOp, "Output Slice : {}->{}", output_slice_height_start, output_slice_height_end);
+                    log_debug(tt::LogOp, "Input Slice : {}->{}", input_slice_height_start, input_slice_height_end);
+                    log_debug(tt::LogOp, "Padding : {}->{}", pad_top, pad_bottom);
+
+                    if(input_slice_height_start < input_slice_height_end) {
+                        auto sliced_input_tensor = ttnn::slice(input_tensor,
```

Comment:
would it be more optimal to use interleaved_to_sharded_partial instead slice/to_layout/concat?

Reply:
I think interleaved_to_sharded_partial is hardcoded for 3D Tensors, but we have 4D Tensors.